Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Vasily Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasily Nesterov
+-/
+import Mathlib.Analysis.Convex.Combination
+import Mathlib.LinearAlgebra.AffineSpace.Independent
+
+/-!
+# Radon's convexity theorem
+
+Radon's theorem on convex sets states that any affine dependent set can be partitioned into two sets
+whose convex hulls intersect.
+
+## Main results
+
+* `Radon_partition`: Radon convexity theorem
+
+## Tags
+convex hull, radon
+
+-/
+
+
+open Set Finset
+open BigOperators
+
+universe u
+
+variable {𝕜 : Type*} {E : Type u} [LinearOrderedField 𝕜] [AddCommGroup E] [Module 𝕜 E]
+
+/-- Any family `f` of affine dependent vectors contains a set `I` with the property that
+convex hulls of `I` and `Iᶜ` intersect. -/
+theorem radon_partition {ι : Type*} {f : ι → E}
+    (h : ¬AffineIndependent 𝕜 f) : ∃ (I : Set ι),
+    (Set.Nonempty ((convexHull 𝕜 (f '' I)) ∩ (convexHull 𝕜 (f '' Iᶜ)))) := by
+  rw [affineIndependent_iff] at h; push_neg at h
+  rcases h with ⟨s1, w, h_wsum, h_vsum, nonzero_w_index, h1, h2⟩
+  let I : Finset ι := s1.filter (fun i ↦ 0 ≤ w i)
+  let J : Finset ι := s1.filter (fun i ↦ w i < 0)
+  use (I : Set ι)
+
+  let weights_sum_I := ∑ i in I, w i
+  let weights_sum_J := ∑ i in J, w i
+  have h3 : weights_sum_I + weights_sum_J = 0 := by
+    simpa only [← h_wsum, not_le] using sum_filter_add_sum_filter_not s1 (fun i ↦ 0 ≤ w i) w
+
+  have h_I_pos : 0 < weights_sum_I := by
+    rcases exists_pos_of_sum_zero_of_exists_nonzero _ h_wsum ⟨nonzero_w_index, h1, h2⟩
+      with ⟨pos_w_index, h1', h2'⟩
+    exact sum_pos'
+      (by simp only [mem_filter, and_imp, imp_self, implies_true])
+      ⟨pos_w_index, by simp only [mem_filter, h1', h2'.le], h2'⟩
+
+  let w' : ι → 𝕜 := fun i ↦ |w i| / weights_sum_I
+  let p : E := ∑ i in I, (w' i) • (f i) -- point of intersection
```

Comment:
You could define this using `Finset.centerMass` which hopefully make everything easier.